content/academy/puppeteer_playwright/collecting_data.md

```diff
@@ -0,0 +1,311 @@
+await page.waitForNetworkIdle();
+
+await autoScroll(page);
+
+// code will go here
+
+await sleep(10);
+
+await browser.close();
+</marked-tab>
+```
+
+Notice that we've written and are using a custom `autoScroll()` function, which will scroll to the very bottom of the page. We are doing this because the images of the products are lazy-loaded (they are only loaded once they've been scrolled into view).
+
+## [](#collecting-in-page-evaluate) Collecting in the browser context
+
+Whatever is returned from the callback function in `page.evaluate()` will be returned by the evaluate function, which means that we can set it to a variable like so:
+
+```JavaScript
+const products = await page.evaluate(() => {});
+```
+
+We'll be returning a bunch of product objects from this function, which will be accessible back in our Node.js context after the promise has resolved. Let's now go ahead and write some data collection code to collect each product:
+
+```JavaScript
+const products = await page.evaluate(() => {
+    return [...document.querySelectorAll('a[class*="ProductCard_root"]')].map((elem) => {
+        const name = elem.querySelector('h3[class*="ProductCard_name"]').textContent;
+        const price = elem.querySelector('div[class*="ProductCard_price"]').textContent;
+        const image = elem.querySelector('img').getAttribute('src');
+
+        return {
+            name,
+            price,
+            image: `https://demo-webstore.apify.org${image}`,
+        };
+    });
+});
+
+console.log(products);
+```
+
+When we run this code, we see this logged to our console:
+
+![Products logged to the console]({{@asset puppeteer_playwright/images/log-products.webp}})
+
+Great! So we're done, right? Well, not quite. Let's throw an error within the function if it includes the keyword **sunglasses**:
+
+```JavaScript
+const products = await page.evaluate(() => {
+    return [...document.querySelectorAll('a[class*="ProductCard_root"]')].map((elem) => {
+        const name = elem.querySelector('h3[class*="ProductCard_name"]').textContent;
+
+        if (name.toLowerCase().includes('sunglasses')) {
+            throw new Error('oops');
+        }
+
+        const price = elem.querySelector('div[class*="ProductCard_price"]').textContent;
+        const image = elem.querySelector('img').getAttribute('src');
+
+        return {
+            name,
+            price,
+            image: `https://demo-webstore.apify.org${image}`,
+        };
+    });
+});
+
+console.log(products);
+```
+
+When we run this code, we get a nasty error that looks like this:
+
+```text
+Error: Evaluation failed: Error: oops
+    at __puppeteer_evaluation_script__:6:19
+    at Array.map (<anonymous>)
+    at __puppeteer_evaluation_script__:2:75
+    ...
+```
+
+So, just because our function is failing to collect one or two of the products, it's failing to collect all of them and completely crashing our program. The solution for this is to parse the data in the Node.js context instead.
```

Comment:
I think this is a bit unfair reason to do the parsing in cheerio. You can handle the errors with a simple try catch. There are other reasons to do it the cheerio way:

- you can easily port the code between headless browsers and plain http scraping
- you don't have to worry in which context you're working
- I don't know, I'm sure something else 😄 

I would use this opportunity to
1. first do it the evaluate + document.querySelector way
2. then show them how to inject jQuery into the browser and use it in there
3. finally show them that the easiest way is probably to extract the whole HTML and do it the cheerio way in Node